sota-client/src/atomic.rs

```diff
@@ -48,34 +37,64 @@ pub trait Bus: Send {
     fn write_message(&self, msg: &Message) -> Result<(), Error>;
 }
 
+/// Define the interface for transitioning a `Secondary` to the next state.
+pub trait Step: Send {
+    fn step(&mut self, state: State, payload: Option<Payload>) -> Result<Option<StepData>, Error>;
+}
+
+
+/// All possible states for a `Primary` or `Secondary`.
+#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum State {
+    Idle,
+    Ready,
+    Verify,
+    Fetch,
+    Commit,
+    Abort,
+}
+
 /// A `Bus` message to be picked up by either a `Primary` or a `Secondary`.
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
 pub enum Message {
-    Next { txid: Uuid, serial: String, state: State, payload: Vec<u8> },
-    Ack  { txid: Uuid, serial: String, state: State, report: Vec<u8> },
+    Next { txid: Uuid, serial: String, state: State, payload: Option<Payload> },
+    Ack  { txid: Uuid, serial: String, state: State, report: Option<TufSigned> },
 
     Req  { txid: Uuid, serial: String, image: String, index: u64 },
     Resp { txid: Uuid, serial: String, image: String, index: u64, chunk: Vec<u8> },
 }
 
+/// A fallback manifest per ECU that will be sent in case of a timeout.
+pub type Manifests = HashMap<String, TufSigned>;
 
-/// Transition a `Secondary` to the next state.
-pub trait Step: Send {
-    fn step(&mut self, state: State, payload: &[u8]) -> Result<StepData, Error>;
-}
+/// A mapping from serials to the payloads to be delivered at each state.
+pub type Payloads = HashMap<String, HashMap<State, Payload>>;
 
-/// An `ImageWriter` should be returned at the `Fetch` state, while a
-/// `TufSigned` report should be returned at a `Commit` or `Abort` state.
-pub struct StepData {
-    pub report: Option<TufSigned>,
-    pub writer: Option<ImageWriter>,
+/// Data that may be delivered to a `Secondary` before executing state transition.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
```

Comment:
Until there are custom bincode serialization and deserialization functions, yes. I'll add those in a separate PR.